File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/finder/PolygonFinder.cpp
/*
 * PolygonFinder.cpp
 *
 *  Created on: 24 nov 2018
 *      Author: ema
 *      Copyright 2025 Emanuele Cesaroni
 */

#include <iostream>
#include <list>
#include <map>
#include <ctime>
#include <typeinfo>
#include <string>
#include <vector>
#include <utility>
#include "PolygonFinder.h"
#include "../bitmaps/Bitmap.h"
#include "../matchers/Matcher.h"
#include "../matchers/RGBMatcher.h"
#include "../matchers/RGBNotMatcher.h"
#include "optionparser.h"
#include "NodeCluster.h"
#include "Node.h"
#include "FinderUtils.h"

PolygonFinder::PolygonFinder(Bitmap *bitmap,
                             Matcher *matcher,
                             Bitmap *test_bitmap,
                             std::vector<std::string> *options,
                             int start_x,
                             int end_x)
    : source_bitmap(bitmap),
      matcher(matcher),
      start_x(start_x),
      end_x(end_x == -1 ? bitmap->w() : end_x)
{ this->rgb_matcher = dynamic_cast<RGBMatcher*>(matcher);
  if (options != nullptr) FinderUtils::sanitize_options(this->options, options);
  this->node_cluster = new NodeCluster(source_bitmap->h(), source_bitmap->w(), &this->options);

  //= SCAN ==============//
  cpu_timer.start();
  scan();
  reports["scan"] = cpu_timer.stop();
  //=====================//

  //= BUILD_TANGS_SEQUENCE ===//
  cpu_timer.start();
  node_cluster->build_tangs_sequence();
  reports["build_tangs_sequence"] = cpu_timer.stop();
  //=====================//

  //= PLOT ===//
  cpu_timer.start();
  node_cluster->plot(this->options.versus);
  reports["plot"] = cpu_timer.stop();
  //=====================//

  //= COMPRESS_COORDS ===//
  cpu_timer.start();
  node_cluster->compress_coords(this->node_cluster->polygons, this->options);
  reports["compress"] = cpu_timer.stop();
  //=====================//
  reports["total"] = reports["scan"] + reports["build_tangs_sequence"] + reports["plot"] + reports["compress"];
}

std::list<ShapeLine*> *PolygonFinder::get_shapelines() {
  std::list<ShapeLine*> *sll = new std::list<ShapeLine*>();
  for (int line = 0; line < this->node_cluster->height; line++) {
    for (Node& node : this->node_cluster->vert_nodes[line]) {
      ShapeLine *sl = new ShapeLine({node.min_x, node.max_x, node.y});
      sll->push_back(sl);
    }
  }
  return sll;
}

PolygonFinder::~PolygonFinder() {
  delete this->node_cluster;
}

void PolygonFinder::scan() {
    int bpp = this->source_bitmap->get_bytes_per_pixel();

    RGBMatcher* rgb_m = dynamic_cast<RGBMatcher*>(this->matcher);

    if (bpp == 1) {
      auto fetcher = [](const unsigned char* p) { return static_cast<unsigned int>(p[0]); };
      if (rgb_m) run_loop(rgb_m, fetcher);
      else       run_loop(this->matcher, fetcher);
    } else {
      auto fetcher = [](const unsigned char* p) { return *reinterpret_cast<const uint32_t*>(p); };
      if (rgb_m) run_loop(rgb_m, fetcher);
      else       run_loop(this->matcher, fetcher);
    }
}

ProcessResult* PolygonFinder::process_info() {
  ProcessResult *pr = new ProcessResult();
  pr->groups = this->node_cluster->sequences.size();
  pr->polygons = std::move(this->node_cluster->polygons);
  pr->benchmarks = std::move(this->reports);
  pr->treemap = this->node_cluster->treemap;

  if (this->node_cluster->options->named_sequences && typeid(*this->source_bitmap) == typeid(Bitmap))
  { std::string sequence;
    int n = 0;
    for (const auto& seq_list : this->node_cluster->sequences) {
      std::string seq;
      for (Node* node : seq_list) {
        seq += node->name;
      }
      if (n != 0) sequence += '-';
      sequence += seq;
      n++;
    }
    pr->named_sequence = sequence;
  }
  else pr->named_sequence = "";
  return(pr);
}

File: lib/contrek/finder/polygon_finder.rb
require "benchmark"
# todo
# - builtin png decoder which makes blocks to pass vert_nodes
# - work only in clockwise (no more :o e :a); finally for coords use push_back on :a and
#   push_front on :o
module Contrek
  module Finder
    class PolygonFinder
      NodeDescriptor = Struct.new(:node, :a, :o)
      def initialize(bitmap, matcher, test_bitmap = nil, options = {})
        @options = {versus: :a}.merge(options)
        sanitize_options
        @source_bitmap = bitmap
        @matcher = matcher

        @test_bitmap = test_bitmap
        @node_cluster = NodeCluster.new(@source_bitmap.h, @options)
        @reports = {}

        # 1 finds matching blocks
        @reports[:scan] = Benchmark.measure do
          scan
        end

        # 2 builds relational spatially tree map
        @reports[:build_tangs_sequence] = Benchmark.measure do
          @node_cluster.build_tangs_sequence
        end

        # 3 plotting
        @reports[:plot] = Benchmark.measure do
          @node_cluster.plot(@test_bitmap)
        end

        # 4 compress
        @reports[:compress] = Benchmark.measure do
          if @options.has_key?(:compress)
            @node_cluster.compress_coords
          end
        end
      end

      def sanitize_options
        @options[:versus] = :a unless @options[:versus] == :a || @options[:versus] == :o
      end

      # infos
      def process_info
        {named_sequence: @node_cluster.sequences.map { |list| list.map(&:name).join }.join("-"),
         groups: @node_cluster.sequences.size,
         groups_names: @node_cluster.root_nodes.map(&:name).join,
         polygons: @node_cluster.polygons,
         benchmarks: format_benchmarks,
         treemap: (@node_cluster.treemap if @options.has_key?(:treemap))}
      end

      def get_shapelines
        shapes = []
        @node_cluster.vert_nodes.each do |line|
          line.each do |node|
            shapes << {start_x: node.min_x, end_x: node.max_x, y: node.y}
          end
        end
        shapes
      end

      def draw_polygons(png_image)
        Contrek::Bitmaps::Painting.direct_draw_polygons(@node_cluster.polygons, png_image)
      end

      def draw_shapelines(png_image)
        slines = get_shapelines
        color = ChunkyPNG::Color("blue @ 1.0")
        slines.each do |sline|
          png_image.draw_line(sline[:start_x], sline[:y], sline[:end_x], sline[:y], color)
        end
      end

      private

      # image scan
      def scan
        last_color = nil
        matching = false
        min_x = 0
        max_x = 0
        @source_bitmap.scan do |x, y, color|
          if @matcher.match?(color) && matching == false
            min_x = x
            last_color = color
            matching = true
            if x == (@source_bitmap.w - 1)
              max_x = x
              Contrek::Finder::Node.new(@node_cluster, min_x, max_x, y, last_color)
              matching = false
            end
          elsif @matcher.unmatch?(color) && matching == true
            max_x = x - 1
            Contrek::Finder::Node.new(@node_cluster, min_x, max_x, y, last_color)
            matching = false
          elsif x == (@source_bitmap.w - 1) && matching == true
            max_x = x
            Contrek::Finder::Node.new(@node_cluster, min_x, max_x, y, last_color)
            matching = false
          end
        end
      end

      def format_benchmarks
        r = {}
        total = 0
        @reports.each do |k, v|
          r[k] = (v.real * 1000).round(3)
          total += v.real * 1000
        end
        r[:total] = total.round(3)
        r
      end
    end
  end
end
